File: 224.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> num, ope;
        int res=0, sign=1, t=0;
        for(auto c : s){
            if(isdigit(c)){
                t = t*10 + c - '0';
            }else{
                res += sign*t;
                t = 0;
                switch (c){
                   case '+':
                     sign = 1;
                     break;
                   case '-':
                     sign = -1;
                     break;
                   case '(':
                     num.push(res);
                     ope.push(sign);
                     res = 0;
                     sign = 1;
                     break;
                   case ')':
                     if(!ope.empty()){
                         res = ope.top()*res + num.top();
                         num.pop();
                         ope.pop();
                     }
                }
            }
        }
        res += sign*t;
        return res;
    }
};

File: golang/224/224.go
package main

var strs []rune

func calculate(s string) int {
	strs = []rune(s)
	return helper()
}

func helper() int {
	stack := make([]int, 0)
	num := 0
	sign := '+'

	for len(strs) > 0 {
		v := strs[0]
		strs = strs[1:] // pop left

		if isDigit(v) {
			num = num*10 + int(v-'0')
		} else if v == '(' {
			num = helper()
		}

		if (!isDigit(v) && v != ' ') || len(strs) == 0 {
			switch sign {
			case '+':
				stack = append(stack, num)

			case '-':
				stack = append(stack, -1*num)

			case '*':
				pre := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				stack = append(stack, pre*num)

			case '/':
				pre := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				stack = append(stack, pre/num)
			}

			sign = v
			num = 0
		}

		if v == ')' {
			break
		}
	}

	res := 0
	for _, v := range stack {
		res += v
	}

	return res
}

func isDigit(s rune) bool {
	return s >= '0' && s <= '9'
}
